### PR TITLE
ready: Use envoy listener to expose endpoint from worker for ambassador

### PR DESCRIPTION
**Pull request by jfrabaute**
/ready endpoint used by emissary is using the admin port (8001 by
default).
This generates a problem during config reloads with large configs as the
admin thread is blocking so the /ready endpoint can be very slow to
answer (in the order of several seconds, even more).

The problem is described in this envoy issue:
https://github.com/envoyproxy/envoy/issues/16425

This change is fixing the slowness of /ready endpoint.
The /ready endpoint is now exposed in the worker pool by adding a
listener+ health check http filter.
This way, the /ready endpoint is fast and it is not blocked by any
config reload or blocking admin operation as it depends on the worker
pool.

## Related Issues
https://github.com/envoyproxy/envoy/issues/16425

## Testing
Manual tests
Deployed on a local k8s cluster.

### CODE REVIEW

**Review by alexgervais — COMMENTED** [review]
Thanks for your contribution @jfrabaute!
I'm a bit curious about your testing strategy for this PR. How did you go about testing large configuration reloads? Could we automate any part of your manual tests?

**Comment by jfrabaute** [comment]
Hi,

For testing, I did test on the previous PR: https://github.com/emissary-ingress/emissary/pull/3626
This PR is the exact same except that it also it also impacts the go process to check the readiness on this new endpoint rather then the old endpoint (which is slow).

In order to test manually, you can start an ambassador instance on a cluster(or a kind cluster for instance), create 1000 mappings.
When don, you can hit the stats endpoint on envoy (something like http://localhost:8001/stats?format=prometheus) and while running this hit, you can query http://localhost:8001/ready right after the stats endpoint has started(but not finished) yet.
You'll see that stats prometheus will take a while, like several seconds, and in the meantime, /ready endpoint will also hang.
Then when the stats endpoint ends, the ready endpoint will exit. So the ready endpoint execution on the envoy side is blocked by the stats endpoint (the admin thread is single threaded).
When moving the ready endpoint to a worker listener, the hit is always fast, just a efw milliseconds, regardless if a stats endpoint execution is in progress or not.

File: pkg/acp/envoy.go [hunk 1]
```diff
@@ -144,3 +151,19 @@ func (w *EnvoyWatcher) IsAlive() bool {
...
+	var err error
+	strReadyPort := os.Getenv("AMBASSADOR_READY_PORT")
+	if strReadyPort != "" {
+		readyPort, err = strconv.Atoi(strReadyPort)
+		if err != nil {
+			dlog.Infof(context.Background(), "Unable to parse AMBASSADOR_READY_PORT: %s", err)
+		}
+	}
+	if readyPort < 1 || readyPort > 32767 {
+		readyPort = 8002
```
[the remark this answers is not in the document]
Comment (alexgervais) [reply pkg/acp/envoy.go:166]: Thanks for the clarification, I understand much better now.
I know a lot of installations and monitors are configured to ping `:8001/ready` for uptime and health check of course. I wonder how we can make a smooth transition to the new port... any thoughts @LanceEa ?

[the remark this answers is not in the document]
Comment (jfrabaute) [reply pkg/acp/envoy.go:166]: `:8001/ready` endpoint is still working. This is an endpoint automatically exposed by the envoy admin thread, so installs using this endpoint should still work.

**Comment by jfrabaute** [comment]
Hi,

Any chance to have this PR reviewed and merged? or get feedback about what's missing for a merge?

Thanks.

**Comment by jfrabaute** [comment]
Superseded by: https://github.com/emissary-ingress/emissary/pull/4300

**Pull request closed without merging**